### ISSUE TITLE
can someone check whether DH / ECC is still broken in current JDK versions?

### ISSUE THREAD

**Issue by calestyo** [dCache/dcache#1769]
Hey.

Since some time you disable DH kex and ECC per default, with the reasons:
>   DISABLE_BROKEN_DH
>
>   Diffie-Hellman is broken in Java 1.7u6 and forward. If  DISABLE_BROKEN_DH is
>   included, dCache will disable all cipher families involving Diffie-Hellman on
>   those versions of Java. Depending on the client, this may result in less
>   secure SSL/TLS/GSI connections.

plus

>   Elliptic Curve ciphers are broken in Java 1.7 on Linux. The problem is that the
>   JRE successfully negotiates the use of cipher variants not supported by libnss3.
>   If this option is specified, dCache will disable all cipher families involving
>   Elliptic Curve ciphers.

As you already noted, this is not the best for security,.. especially since ECC is disabled as well, there will be neither DH nor ECDH and thus no PFS TLS.

I did a very quick search on Google (not going too much into depth so) and there were some hints that implied that this could have been fixed in current Java versions.

Could someone check this up?

Thanks,
Chris.

**Comment by paulmillar** [comment]
The underlying problem that DISABLE_BROKEN_DH works-around does not affect Java 8.  Therefore, the flag only has an effect when running dCache with Java 7.  The underlying problem is first present from 7u5 and is fixed in 7u51.  I'm working on a patch that will switch off DISABLE_BROKEN_DH if dCache is run with 7u51 or newer.

DISABLE_EC is more problematic: the underlying cause comes from a platform-integration issue; specifically, with the limited support for ECDH in libnss3.  As it's a platform-specific issue, there's little we can do (or really, little we want to get involved with) within dCache as we want to remain platform agnostic.

That said, I'm adding some documentation that describes how to check if your OpenJDK is affected.  I think we'll need to keep DISABLE_EC enabled by default for the time being, but sites that care can verify whether they have working ECDH and remove the flag.

**Comment by calestyo** [comment]
great... is the check for the ECDH already available? Cause then I could already enable it again during the current downtime :-)

**Comment by calestyo** [comment]
See #1768.

**Comment by gbehrmann** [comment]
For 2.12 and 2.13 we can simply drop all support for DISABLE_BROKEN_DH as Java 8 is required and it is not possible to run dCache on a JVM with broken DH support.

**Comment by paulmillar** [comment]
DISABLE_BROKEN_DH is dropped in master -- I decided against dropping it in 2.13 as someone could have modified the configuration (e.g., to add DISABLE_RC4), which would cause dCache to fail after upgrade.  In any case, DISABLE_BROKEN_DH does nothing so is not dangerous.

I've added a description of how to test whether DISABLE_EC is needed to the properties file.  The problem is platform-dependent.  In RedHat/SL they solved the problem by default configuration banning ECDH; Debian appear to have fixed the underlying library or added support for a provider providing only partial ECDH support.  Therefore we should keep DISABLE_EC present for now; sites can verify if their java supports ECDH and remove the flag as appropriate.
